subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java

```diff
@@ -186,7 +187,7 @@ public void setBuildName(String buildName) {
     @TaskAction
     void build() {
         // TODO: Allow us to inject plugins into GradleBuild nested builds too.
-        BuildDefinition buildDefinition = BuildDefinition.fromStartParameter(getStartParameter(), getServices().get(PublicBuildPath.class));
+        BuildDefinition buildDefinition = BuildDefinition.fromStartParameter((StartParameterInternal) getStartParameter(), getServices().get(PublicBuildPath.class));
```

Comment:
This is now a potential `ClassCastException` as a user can call `new StartParameter()` to create a new instance.

Taking a deeper look it seems that we'd already get the exception if an included build was used, but not when included builds are not used. So I guess this is really a new error we are introducing here.

I actually have no good idea how to solve this. :/

One way would be to deprecate the usage of `setStartParameter()`, and do some magic to wrap the value into some sort of delegate for now. I'm not sure it's worth it tbh.

Reply:
You will get an exception if you don't use `StartParameterInternal` to run the build. The last moment when that happens is when the provided start parameter is cast to `StartParameterInternal` which happens for example when we create the `buildSrc`-classpath: https://github.com/gradle/gradle/blob/9f988597014407d1569fd81be97cb6651ac785f2/subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java#L90

So I don't think casting to `StartParameterInternal` is a breaking change.

Reply:
OK. Should we then deprecate `new StartParameter()` and maybe `GradleBuild.setStartParameter()`?